### PR TITLE
Rename TyVariants and variants

### PR DESCRIPTION
**Pull request by varkor**
- Rename `TypeVariants` to `TyKind`.
- Remove the `Ty` prefix from each one of its variants (plus the identically-named variants of `PrimTy`).
- Rename `ty::Slice` to `ty::List`.

The new names look cleaner.

r? @eddyb

### CODE REVIEW

File: src/librustc/hir/def.rs [hunk 1]
```diff
@@ -53,13 +53,13 @@ pub enum Def {
     Existential(DefId),
     /// `type Foo = Bar;`
     TyAlias(DefId),
-    TyForeign(DefId),
+    Foreign(DefId),
     TraitAlias(DefId),
     AssociatedTy(DefId),
     /// `existential type Foo: Bar;`
     AssociatedExistential(DefId),
     PrimTy(hir::PrimTy),
```
Comment (Mark-Simulacrum) [inline src/librustc/hir/def.rs]: Hm, maybe drop `Ty` here too? And in the other cases perhaps?

**Comment by eddyb** [comment]
cc @nikomatsakis

File: src/librustc/hir/def.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc/hir/def.rs]: No, this is "primitive type". `Def` contains both types and values.

File: src/librustc/hir/def.rs [hunk 2]
```diff
@@ -53,13 +53,13 @@ pub enum Def {
     Existential(DefId),
     /// `type Foo = Bar;`
     TyAlias(DefId),
-    TyForeign(DefId),
+    Foreign(DefId),
```
Comment (eddyb) [inline src/librustc/hir/def.rs]: This seems suboptimal... what if you *first* rename this to `ForeignTy` *and then* do the other renames?

File: src/librustc/hir/def.rs [hunk 3]
```diff
@@ -53,13 +53,13 @@ pub enum Def {
...
     TraitAlias(DefId),
     AssociatedTy(DefId),
     /// `existential type Foo: Bar;`
     AssociatedExistential(DefId),
     PrimTy(hir::PrimTy),
-    TyParam(DefId),
+    Param(DefId),
```
Comment (eddyb) [inline src/librustc/hir/def.rs]: We probably want to keep this the way it is, assuming we need to tell apart ty params from const ones.

File: src/librustc/hir/def.rs [hunk 4]
```diff
@@ -53,13 +53,13 @@ pub enum Def {
     Existential(DefId),
     /// `type Foo = Bar;`
     TyAlias(DefId),
-    Foreign(DefId),
+    TyForeign(DefId),
```
Comment (eddyb) [inline src/librustc/hir/def.rs]: I prefer `ForeignTy`, specifically, just like `AssociatedTy` below.

File: src/test/ui/privacy/private-inferred-type-1.rs [hunk 5]
```diff
@@ -11,15 +11,15 @@
...
+trait Param {
     fn ty_param_secret(&self);
 }
 
 mod m {
     struct Priv;
 
     impl ::Arr0 for [Priv; 0] { fn arr0_secret(&self) {} }
-    impl ::TyParam for Option<Priv> { fn ty_param_secret(&self) {} }
+    impl ::Param for Option<Priv> { fn ty_param_secret(&self) {} }
```
Comment (eddyb) [inline src/test/ui/privacy/private-inferred-type-1.rs]: Hehe you don't need to change this.

Comment (varkor) [reply src/test/ui/privacy/private-inferred-type-1.rs]: Fixed.

**Comment by eddyb** [comment]
@varkor r=me if you rebase

**Comment by varkor** [comment]
@bors r=eddyb p=10

(This bitrots very quickly.)

**Pull request merged**